cocos2d/core/renderer/RendererWebGL.js

```diff
@@ -147,8 +219,439 @@ cc.rendererWebGL = {
+
+        gl.bufferSubData(gl.ELEMENT_ARRAY_BUFFER, 0, indices);
+
+        if (!CACHING_BUFFER) {
+            _batchBufferPool.put(buffer);
+        }
+        return count;
+    },
+
+    // Auto batch implementation inspired from @Heishe 's PR
+    // Ref: https://github.com/cocos2d/cocos2d-html5/pull/3248
+    // Batch rendering using result collected in `_forwardBatch`
+    _batchRendering: function () {
+        // var node = this._node;
+        var texture = _batchedInfo.texture;
+        var shader = _batchedInfo.shader;
+        var count = _currentBuffer.count;
+
+        var bytesPerRow = 16; //4 floats with 4 bytes each
+
+        shader.use();
+        shader._updateProjectionUniform();
+
+        cc.glBlendFunc(_batchedInfo.blendSrc, _batchedInfo.blendDst);
+        cc.glBindTexture2DN(0, texture);                   // = cc.glBindTexture2D(texture);
+
+        gl.bindBuffer(gl.ARRAY_BUFFER, _currentBuffer.buffer.arrayBuffer);
+
+        cc.glEnableVertexAttribs(cc.VERTEX_ATTRIB_FLAG_POS_COLOR_TEX);
+
+        var vertexOffset = _currentBuffer.vertexOffset;
+        gl.vertexAttribPointer(0, 3, gl.FLOAT, false, 24, vertexOffset);
+        gl.vertexAttribPointer(1, 4, gl.UNSIGNED_BYTE, true, 24, vertexOffset + 12);
+        gl.vertexAttribPointer(2, 2, gl.FLOAT, false, 24, vertexOffset + 16);
+        
+        var i;
+        var matrixOffset = _currentBuffer.matrixOrigin + _currentBuffer.matrixOffset;
+        // Enable matrix vertex attribs row by row (vec4 * 4)
+        for (i = 0; i < 4; ++i) {
+            gl.enableVertexAttribArray(cc.VERTEX_ATTRIB_MVMAT0 + i);
+            gl.vertexAttribPointer(cc.VERTEX_ATTRIB_MVMAT0 + i, 4, gl.FLOAT, false, bytesPerRow * 4, matrixOffset + bytesPerRow * i); //stride is one row
+        }
+
+        gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, _currentBuffer.buffer.elementBuffer);
+        gl.drawElements(gl.TRIANGLES, _currentBuffer.totalIndexSize, gl.UNSIGNED_SHORT, _currentBuffer.indexOffset);
+
+        for (i = 0; i < 4; ++i) {
+            gl.disableVertexAttribArray(cc.VERTEX_ATTRIB_MVMAT0 + i);
+        }
+
+        cc.g_NumberOfDraws++;
+    },
+
+    /**
+     * drawing all renderer command to context (default is cc._renderContext)
+     * @param {WebGLRenderingContext} [ctx=cc._renderContext]
+     */
+    rendering: function (ctx) {
+        var locCmds = this._renderCmds,
+            i, len, cmd, next, batchCount,
+            context = ctx || cc._renderContext;
+
+        // Only update virtual buffers if children order dirty in the current frame
+        if (ACTIVATE_AUTO_BATCH && (_orderDirtyInFrame || _bufferError)) {
+            this._refreshVirtualBuffers();
+        }
+
+        for (i = 0, len = locCmds.length; i < len; i++) {
+            cmd = locCmds[i];
+            next = locCmds[i+1];
+
+            if (ACTIVATE_AUTO_BATCH) {
+                // Already batched in buffer
+                if (cmd._vBuffer) {
+                    batchCount = this._forwardCheck(i);
```

Comment:
Do we really need `_forwardCheck` after calling of `_refreshVirtualBuffers`?
My be add also checking of `_orderDirtyInFrame || _bufferError` here.